src/libbacktrace/fileline.c

```diff
@@ -38,13 +38,40 @@ POSSIBILITY OF SUCH DAMAGE.  */
 #include <fcntl.h>
 #include <stdlib.h>
 
+#define USE_WIN_EXECNAME (defined(_WIN32_WINNT) && (_WIN32_WINNT >= 0x0600))
+#if USE_WIN_EXECNAME
+#include <windows.h>
+#endif /* USE_WIN_EXECNAME */
+
 #include "backtrace.h"
 #include "internal.h"
 
 #ifndef HAVE_GETEXECNAME
 #define getexecname() NULL
 #endif
 
+/* Targeted fix for Rust backtraces on Windows/GNU.
+   We can't set the executable name once during during creation
+   of backtrace state for security reasons, see issue #21889 for
+   details. So we recalculate it each time the file is accessed.
+   We also use QueryFullProcessImageName instead of APIs like
+   GetModuleFileName because it's correctly updated when the
+   executable is renamed. */
+#if USE_WIN_EXECNAME
+#undef getexecname
+const char *getexecname(void) {
+  /* Accesses to backtrace functionality from Rust are serialized,
+     so having a single static unsyncronized buffer is enough. */
+  static char buf[MAX_PATH];
+  /* The returned name is later passed to `open`, so it needs to
+     be encoded in the current locale, so we use `QueryFullProcessImageNameA`.
+     As a result paths not representable in the current locale are not supported. */
```

Comment:
This is super unfortunate.
Yet another reason to avoid MinGW and friends on Windows.